src/libstd/Cargo.toml

```diff
@@ -48,3 +48,4 @@ jemalloc = ["alloc_jemalloc"]
 force_alloc_system = []
 panic-unwind = ["panic_unwind"]
 profiler = ["profiler_builtins"]
+wasm_syscall = []
```

Comment:
So I guess this means the `std` will have to be recompiled to use this? ):

To be honest I would *really* prefer another way of opting into this than having to recompile the whole `std`. I really don't want to have to add support for `std` recompilation to `cargo-web` on top of the gazzilion of hoops I have to go through to get something usable. ):

Reply:
It looks like it will be fairly easy to use Xargo to build libstd with the additional feature, but I don't now how that will inter-operate with `cargo-web`.

If you find an alternative way to dynamically configure whether or not `libstd` imports a symbol, then please suggest it, but the way this is done currently for things like the allocator/panic selection is a hack, and @alexcrichton was reluctant to build on this hack for something specific to wasm.

Reply:
> If you find an alternative way to dynamically configure whether or not `libstd` imports a symbol, then please suggest it, but the way this is done currently for things like the allocator/panic selection is a hack, and @alexcrichton was reluctant to build on this hack for something specific to wasm.

As far as I understand the appoach used for allocators (etc.) is something like this in a nutshell: in `std` you put an `extern` for a function, and then at link time you have two crates which implement that symbol and you link to one of them based on what the user wants. Correct? And this is the approach you wanted to originally take (but with only one crate instead of two, to force the generation of an import.)

To me *personally* this doesn't really feel like a hack, and seems like the simplest way to do it. Anything I can think of is going to be more complex than that, so I guess if Alex is against it then that's it. ):

So if we're going to oficially require a recompilation of the `std` here's a request from me - @alexcrichton could we *at least* define the empty syscall wrapper inside of `std` like this?

Before:

```rust
#[cfg(not(wasm_syscall))]
unsafe fn syscall<T>(_index: SysCallIndex, _data: &mut T) -> bool {
    false
}
```

After:

```rust
#[cfg(not(wasm_syscall))]
#[allow(private_no_mangle_fns)]
#[no_mangle]
#[inline(never)]
unsafe fn __rust_wasm_syscall(_index: SysCallIndex, _data: *mut c_void) -> bool {
    asm!(""); // To prevent inlining since #[inline(never)] in not enough on the wasm32-unknown-unknown target.
    false
}
```

Basically, make it not inlineable and give it a unique name. This puts zero maintenance burden on Rust developers and will allow me to trivially support this in `cargo-web` without having to recompile the `std`.

Reply:
My personal feeling is that we should explicitly allow people to compile their own wasm stds easily - this PR is deliberately minimal and it's not hard to imagine people wanting to hook up additional bits and pieces (e.g. FS calls to a virtual filesystem in JS, tcp sockets to websockets). But to do this we need to make it dead easy to compile these different stds for different use-cases.

I'd personally prefer all the changes to the shim and to libstd to end up in a third party repository rather than being landed here so that it will (as I put it in my previous comment) set the scene for other people to get involved.

I am looking at Xargo at the moment and will continue tomorrow - potentially this shouldn't require much extra effort from cargo web, since the heavy lifting should be done by Xargo.

Reply:
> Basically, make it not inlineable and give it a unique name. This puts zero maintenance burden on Rust developers and will allow me to trivially support this in cargo-web without having to recompile the std.

Can you elaborate what this would do/how it would help you? If nothing else, the answer should probably be in a comment for the function :)

Reply:
> My personal feeling is that we should explicitly allow people to compile their own wasm stds easily

I wouldn't really mind that much if it was oficially supported by cargo/rust (e.g. you could just put the `std` in your `[dependencies]` section), but unfortunately it's not yet.

> Can you elaborate what this would do/how it would help you? If nothing else, the answer should probably be in a comment for the function :)

Sure. Currently to support the `js!` macro from [stdweb](https://github.com/koute/stdweb) I'm already doing a bunch of [transformations on the WebAssembly bytecode](https://github.com/koute/cargo-web/blob/master/src/wasm.rs) which Rust generates. One of the things I can do with my current infrastructure is to freely remap the functions as I see fit and modify the import tables, so *if* I'll be able to identify *which* exact function in the bytecode is the syscall wrapper then I can just replace it with an import in something like ~15 lines of code. Rust itself can keep the status quo. Everyone's happy.